File: src/services/authService.ts
import {
  signInWithPopup,
  GoogleAuthProvider,
  signOut as firebaseSignOut,
  onAuthStateChanged,
  User,
  createUserWithEmailAndPassword,
  signInWithEmailAndPassword,
} from 'firebase/auth';
import {
  doc,
  setDoc,
  getDoc,
  updateDoc,
  serverTimestamp,
  collection,
  Timestamp,
} from 'firebase/firestore';
import { auth, db } from '../config/firebase';
import { UserProfile, UserPreferences, UserData, MemoryData } from '../types/user';
import { FamilyMember } from '../types/calendar';

const googleProvider = new GoogleAuthProvider();
googleProvider.addScope('https://www.googleapis.com/auth/calendar');
googleProvider.addScope('https://www.googleapis.com/auth/userinfo.profile');
googleProvider.addScope('https://www.googleapis.com/auth/userinfo.email');
googleProvider.addScope('profile');
googleProvider.addScope('email');

// Initialize default user preferences
const getDefaultPreferences = (): UserPreferences => ({
  theme: 'system',
  language: 'en',
  timezone: Intl.DateTimeFormat().resolvedOptions().timeZone,
  calendarView: 'week',
  notifications: true,
});

// REMOVED: Default family members are now managed through family system
// REMOVED: Default memory data is now managed through family system

// Create user profile in Firestore
const createUserProfile = async (user: User): Promise<UserProfile> => {
  const userProfile: UserProfile = {
    uid: user.uid,
    email: user.email || '',
    displayName: user.displayName || user.email?.split('@')[0] || 'User',
    photoURL: user.photoURL || undefined,
    phoneNumber: user.phoneNumber || undefined,
    createdAt: new Date(),
    lastLoginAt: new Date(),
    preferences: getDefaultPreferences(),
    // familyId will be set when user creates or joins a family
  };

  // Prepare data for Firestore (remove undefined values)
  const firestoreData: any = {
    uid: user.uid,
    email: user.email || '',
    displayName: user.displayName || user.email?.split('@')[0] || 'User',
    preferences: getDefaultPreferences(),
    createdAt: serverTimestamp(),
    lastLoginAt: serverTimestamp(),
  };

  // Only add optional fields if they exist
  if (user.photoURL) {
    firestoreData.photoURL = user.photoURL;
  }
  if (user.phoneNumber) {
    firestoreData.phoneNumber = user.phoneNumber;
  }

  // Create user document
  await setDoc(doc(db, 'users', user.uid), firestoreData);

  // Create empty credentials document
  await setDoc(doc(db, 'users', user.uid, 'data', 'credentials'), {});

  return userProfile;
};

// Get or create user profile
const getUserProfile = async (user: User): Promise<UserProfile> => {
  const userDoc = await getDoc(doc(db, 'users', user.uid));

  if (userDoc.exists()) {
    // Update last login
    await updateDoc(doc(db, 'users', user.uid), {
      lastLoginAt: serverTimestamp(),
    });

    const data = userDoc.data();
    return {
      ...data,
      createdAt: data.createdAt?.toDate() || new Date(),
      lastLoginAt: new Date(),
    } as UserProfile;
  }

  // Create new profile
  return await createUserProfile(user);
};

// Sign in with Google
export const signInWithGoogle = async (): Promise<UserProfile> => {
  try {
    const result = await signInWithPopup(auth, googleProvider);
    const user = result.user;

    // Debug: Log user data
    console.log('Google Sign-In User Data:', {
      displayName: user.displayName,
      email: user.email,
      photoURL: user.photoURL,
      providerId: user.providerData[0]?.providerId,
    });

    // Get user profile (creates if doesn't exist)
    const profile = await getUserProfile(user);

    // Get Google Calendar credentials if available
    const credential = GoogleAuthProvider.credentialFromResult(result);
    const token = credential?.accessToken;

    if (token) {
      // Store Google Calendar token (update existing credentials doc)
      try {
        const credentialsRef = doc(db, 'users', user.uid, 'data', 'credentials');
        const credentialsDoc = await getDoc(credentialsRef);
        
        if (credentialsDoc.exists()) {
          await updateDoc(credentialsRef, {
            googleCalendarToken: token,
          });
        } else {
          await setDoc(credentialsRef, {
            googleCalendarToken: token,
          });
        }
      } catch (error) {
        console.error('Error storing Google Calendar token:', error);
        // Don't fail the sign-in if token storage fails
      }
    }

    return profile;
  } catch (error: any) {
    console.error('Error signing in with Google:', error);
    throw new Error(error.message || 'Failed to sign in with Google');
  }
};

// Sign up with email and password
export const signUpWithEmail = async (
  email: string,
  password: string,
  displayName: string
): Promise<UserProfile> => {
  try {
    const result = await createUserWithEmailAndPassword(auth, email, password);
    const user = result.user;

    // Update display name
    const profile = await createUserProfile(user);
    await updateDoc(doc(db, 'users', user.uid), {
      displayName,
    });

    return { ...profile, displayName };
  } catch (error: any) {
    console.error('Error signing up with email:', error);
    throw new Error(error.message || 'Failed to sign up');
  }
};

// Sign in with email and password
export const signInWithEmail = async (
  email: string,
  password: string
): Promise<UserProfile> => {
  try {
    const result = await signInWithEmailAndPassword(auth, email, password);
    return await getUserProfile(result.user);
  } catch (error: any) {
    console.error('Error signing in with email:', error);
    throw new Error(error.message || 'Failed to sign in');
  }
};

// Sign out
export const signOut = async (): Promise<void> => {
  try {
    await firebaseSignOut(auth);
  } catch (error: any) {
    console.error('Error signing out:', error);
    throw new Error(error.message || 'Failed to sign out');
  }
};

// Listen to auth state changes
export const onAuthStateChange = (
  callback: (user: UserProfile | null) => void
): (() => void) => {
  return onAuthStateChanged(auth, async (user) => {
    if (user) {
      const profile = await getUserProfile(user);
      callback(profile);
    } else {
      callback(null);
    }
  });
};

// Get current user profile
export const getCurrentUser = async (): Promise<UserProfile | null> => {
  const user = auth.currentUser;
  if (!user) return null;
  return await getUserProfile(user);
};

// Update user profile
export const updateUserProfile = async (
  uid: string,
  updates: Partial<UserProfile>
): Promise<void> => {
  try {
    await updateDoc(doc(db, 'users', uid), {
      ...updates,
      lastLoginAt: serverTimestamp(),
    });
  } catch (error: any) {
    console.error('Error updating user profile:', error);
    throw new Error(error.message || 'Failed to update profile');
  }
};

// Store face descriptor for face recognition
export const storeFaceDescriptor = async (
  uid: string,
  descriptor: number[]
): Promise<void> => {
  try {
    await updateDoc(doc(db, 'users', uid), {
      faceDescriptor: descriptor,
      faceDescriptorUpdatedAt: serverTimestamp(),
    });
  } catch (error: any) {
    console.error('Error storing face descriptor:', error);
    throw new Error(error.message || 'Failed to store face data');
  }
};

// Get user's face descriptor
export const getFaceDescriptor = async (
  uid: string
): Promise<number[] | null> => {
  try {
    const userDoc = await getDoc(doc(db, 'users', uid));
    if (userDoc.exists()) {
      return userDoc.data().faceDescriptor || null;
    }
    return null;
  } catch (error: any) {
    console.error('Error getting face descriptor:', error);
    return null;
  }
};

File: flutter/lib/services/auth_service.dart
import 'package:firebase_auth/firebase_auth.dart';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:google_sign_in/google_sign_in.dart';
import 'firebase_service.dart';

/// User profile data matching web app schema
class UserProfile {
  final String uid;
  final String email;
  final String displayName;
  final String? photoURL;
  final String? phoneNumber;
  final DateTime createdAt;
  final DateTime lastLoginAt;
  final String? familyId;
  final Map<String, dynamic> preferences;

  UserProfile({
    required this.uid,
    required this.email,
    required this.displayName,
    this.photoURL,
    this.phoneNumber,
    required this.createdAt,
    required this.lastLoginAt,
    this.familyId,
    Map<String, dynamic>? preferences,
  }) : preferences = preferences ?? _defaultPreferences();

  static Map<String, dynamic> _defaultPreferences() => {
        'theme': 'system',
        'language': 'en',
        'timezone': DateTime.now().timeZoneName,
        'calendarView': 'week',
        'notifications': true,
      };

  factory UserProfile.fromJson(Map<String, dynamic> json) {
    return UserProfile(
      uid: json['uid'] as String,
      email: json['email'] as String,
      displayName: json['displayName'] as String,
      photoURL: json['photoURL'] as String?,
      phoneNumber: json['phoneNumber'] as String?,
      createdAt: (json['createdAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
      lastLoginAt: (json['lastLoginAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
      familyId: json['familyId'] as String?,
      preferences: json['preferences'] as Map<String, dynamic>? ?? _defaultPreferences(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'uid': uid,
      'email': email,
      'displayName': displayName,
      if (photoURL != null) 'photoURL': photoURL,
      if (phoneNumber != null) 'phoneNumber': phoneNumber,
      'preferences': preferences,
      if (familyId != null) 'familyId': familyId,
    };
  }
}

/// Authentication service for Calendar AI Flutter app.
/// 
/// Provides email/password and Google Sign-In authentication
/// matching the web app's authentication flow.
class AuthService {
  final FirebaseAuth _auth = FirebaseService.auth;
  final FirebaseFirestore _firestore = FirebaseService.firestore;
  final GoogleSignIn _googleSignIn = GoogleSignIn(
    scopes: [
      'email',
      'profile',
      'https://www.googleapis.com/auth/calendar',
    ],
  );

  /// Stream of authentication state changes
  Stream<User?> get authStateChanges => _auth.authStateChanges();

  /// Get current Firebase user
  User? get currentUser => _auth.currentUser;

  /// Check if user is authenticated
  bool get isAuthenticated => currentUser != null;

  /// Sign in with Google
  Future<UserProfile> signInWithGoogle() async {
    try {
      // Trigger Google Sign-In flow
      final GoogleSignInAccount? googleUser = await _googleSignIn.signIn();
      
      if (googleUser == null) {
        throw Exception('Google Sign-In was cancelled');
      }

      // Obtain auth details
      final GoogleSignInAuthentication googleAuth = await googleUser.authentication;

      // Create Firebase credential
      final credential = GoogleAuthProvider.credential(
        accessToken: googleAuth.accessToken,
        idToken: googleAuth.idToken,
      );

      // Sign in to Firebase
      final userCredential = await _auth.signInWithCredential(credential);
      final user = userCredential.user!;

      // Store Google Calendar token if available
      if (googleAuth.accessToken != null) {
        await _storeGoogleCalendarToken(user.uid, googleAuth.accessToken!);
      }

      // Get or create user profile
      return await _getOrCreateUserProfile(user);
    } catch (e) {
      print('Error signing in with Google: $e');
      throw Exception('Failed to sign in with Google: $e');
    }
  }

  /// Sign in with email and password
  Future<UserProfile> signInWithEmail(String email, String password) async {
    try {
      final userCredential = await _auth.signInWithEmailAndPassword(
        email: email,
        password: password,
      );
      
      return await _getUserProfile(userCredential.user!.uid);
    } catch (e) {
      print('Error signing in with email: $e');
      throw Exception('Failed to sign in: $e');
    }
  }

  /// Sign up with email and password
  Future<UserProfile> signUpWithEmail(
    String email,
    String password,
    String displayName,
  ) async {
    try {
      final userCredential = await _auth.createUserWithEmailAndPassword(
        email: email,
        password: password,
      );
      
      final user = userCredential.user!;
      
      // Create user profile with display name
      final profile = await _createUserProfile(user, displayName: displayName);
      
      return profile;
    } catch (e) {
      print('Error signing up: $e');
      throw Exception('Failed to sign up: $e');
    }
  }

  /// Sign out
  Future<void> signOut() async {
    try {
      await Future.wait([
        _auth.signOut(),
        _googleSignIn.signOut(),
      ]);
    } catch (e) {
      print('Error signing out: $e');
      throw Exception('Failed to sign out: $e');
    }
  }

  /// Get user profile from Firestore
  Future<UserProfile> _getUserProfile(String uid) async {
    final doc = await _firestore.collection('users').doc(uid).get();
    
    if (!doc.exists) {
      throw Exception('User profile not found');
    }
    
    // Update last login
    await _firestore.collection('users').doc(uid).update({
      'lastLoginAt': FieldValue.serverTimestamp(),
    });
    
    return UserProfile.fromJson({...doc.data()!, 'uid': uid});
  }

  /// Get or create user profile
  Future<UserProfile> _getOrCreateUserProfile(User user) async {
    try {
      return await _getUserProfile(user.uid);
    } catch (e) {
      // Profile doesn't exist, create it
      return await _createUserProfile(user);
    }
  }

  /// Create new user profile in Firestore
  Future<UserProfile> _createUserProfile(
    User user, {
    String? displayName,
  }) async {
    final profile = UserProfile(
      uid: user.uid,
      email: user.email ?? '',
      displayName: displayName ?? user.displayName ?? user.email?.split('@')[0] ?? 'User',
      photoURL: user.photoURL,
      phoneNumber: user.phoneNumber,
      createdAt: DateTime.now(),
      lastLoginAt: DateTime.now(),
    );

    final data = profile.toJson();
    data['createdAt'] = FieldValue.serverTimestamp();
    data['lastLoginAt'] = FieldValue.serverTimestamp();

    await _firestore.collection('users').doc(user.uid).set(data);
    
    // Create empty credentials document
    await _firestore
        .collection('users')
        .doc(user.uid)
        .collection('data')
        .doc('credentials')
        .set({});

    return profile;
  }

  /// Store Google Calendar token
  Future<void> _storeGoogleCalendarToken(String uid, String token) async {
    try {
      await _firestore
          .collection('users')
          .doc(uid)
          .collection('data')
          .doc('credentials')
          .set({
        'googleCalendarToken': token,
      }, SetOptions(merge: true));
    } catch (e) {
      print('Error storing Google Calendar token: $e');
      // Don't fail the sign-in if token storage fails
    }
  }

  /// Get current user profile
  Future<UserProfile?> getCurrentUserProfile() async {
    final user = currentUser;
    if (user == null) return null;
    
    try {
      return await _getUserProfile(user.uid);
    } catch (e) {
      print('Error getting current user profile: $e');
      return null;
    }
  }

  /// Update user profile
  Future<void> updateUserProfile(
    String uid,
    Map<String, dynamic> updates,
  ) async {
    try {
      await _firestore.collection('users').doc(uid).update({
        ...updates,
        'lastLoginAt': FieldValue.serverTimestamp(),
      });
    } catch (e) {
      print('Error updating user profile: $e');
      throw Exception('Failed to update profile: $e');
    }
  }

  /// Reset password
  Future<void> resetPassword(String email) async {
    try {
      await _auth.sendPasswordResetEmail(email: email);
    } catch (e) {
      print('Error sending password reset email: $e');
      throw Exception('Failed to send password reset email: $e');
    }
  }
}
